pep-0669.rst

```diff
@@ -0,0 +1,318 @@
+mechanism for instrumentation and monitoring that was not possible in 3.10
+or earlier.
+
+Specification
+=============
+
+There are two parts to this specification, instrumentation and monitoring.
+
+Instrumentation occurs early in a program's life cycle and persists through
+the lifetime of the program. It is expected to be pervasive, but fixed.
+Instrumentation is designed to support profiling and coverage tools that 
+expect to be active for the entire lifetime of the program.
+
+Monitoring can occur at any point in the program's life and be applied
+anywhere in the program. Monitoring points are expected to few.
+The capabilities of monitoring are a superset of that of profiling,
+but bulk insertion of monitoring points will be *much* more
+expensive than insertion of instrumentation.
+
+Both instrumentation and monitoring is performed by insertion of
+checkpoints in a code object.
+
+Checkpoints
+-----------
+
+A checkpoint is simply a point in code defined by a 
+``(codeobject, offset)`` pair.
+Every time a checkpoint is reached, the registered callable is called.
+
+Instrumentation
+---------------
+
+Instrumentation supports the bulk insertion of checkpoints, but does not
+allow insertion or removal of checkpoints after code has started to execute.
+
+The events are::
+
+  * BRANCH: A conditional branch is reached. 
+  * JUMPBACK: A backwards, unconditional branch is reached.
+  * ENTER: A Python function is entered.
+  * EXIT: A Python function exits normally (without an exception).
+  * UNWIND: A Python function exits with an unhandled exception.
+  * C_CALL: A call to any object that is not a Python function.
+  * C_RETURN: A return from any object that is not a Python function.
+  * RAISE: An exception is raised.
+  * EXCEPT: An exception is handled.
+
+For each ``ENTER`` event there will be a corresponding
+``EXIT`` or ``UNWIND`` event.
+For each ``C_CALL`` event there will be a corresponding
+``C_RETURN`` or ``RAISE`` event.
+
+All events are integer powers of two and can be bitwise or-ed together to
+instrument multiple events.
+
+Instrumenting code objects
+''''''''''''''''''''''''''
+
+Code objects can be instrumented by calling::
+
+  instrumentation.instrument(codeobject, events)
+
+Code objects must be instrumented before they are executed.
+An exception will be raised if the code object has been executed before it
+is instrumented.
+
+Register callback functions for instrumentation
+'''''''''''''''''''''''''''''''''''''''''''''''
+
+To register a callable for events call::
+
+  instrumentation.register(event, func)
+
+Functions can be unregistered by calling
+``instrumentation.register(event, None)``.
+
+Callback functions can be registered at any time.
+
+Callback function arguments
+'''''''''''''''''''''''''''
+
+When an event occurs the registered function will be called.
+The arguments provided are as follows:
+
+* BRANCH: ``func(code: CodeType, offset: int, taken:bool)``
```

Comment:
Is the code object an important parameter? The frame is generally more useful (and implicitly includes the code object).

Reply:
Then again, I guess for a profiler the code object is enough as it's likely maintaining its own call stacks. Debuggers will need to walk the frame even if callers weren't instrumented, and so having them do an extra call to start walking the stack is better than preemptively passing it in.

If that's the case, a sentence explaining it would be nice, so the next reader doesn't have to guess.

Reply:
+1 to the case where the frame is needed for debuggers (i.e.: this PEP should have some note stating the recommended way to get the frame related to the instrumentation notification).

-- as a note, if performance-wise it'd be the same, I'd say that receiving the frame which contains the code object would be better. If there's some performance penalty, then another way to get it would also be reasonable since it's not always needed (but the recommended way to get it should be explicit in the PEP).